docs/spec/overload.rst

```diff
@@ -99,8 +109,396 @@ variable is not sufficient.
+expanded:
+
+1. Unions: Each subtype of the union should be considered as a separate
+argument type. For example, the type ``int | str`` should be expanded
+into ``int`` and ``str``.
+
+2. ``bool`` should be expanded into ``Literal[True]`` and ``Literal[False]``.
+
+3. ``Enum`` types (other than those that derive from ``enum.Flag``) should
+be expanded into their literal members.
+
+4. ``type[A | B]`` should be expanded into ``type[A]`` and ``type[B]``.
+
+5. Tuples of known length that contain expandable types should be expanded
+into all possible combinations of their subtypes. For example, the type
+``tuple[int | str, bool]`` should be expanded into ``(int, Literal[True])``,
+``(int, Literal[False])``, ``(str, Literal[True])``, and
+``(str, Literal[False])``.
+
+
+[Eric's note for reviewers: I'm not 100% convinced we should
+support argument expansion in all of these cases. Tuple expansion,
+in particular, can be very costly and can quickly blow up in complexity.
+Currently, pyright and mypy support only the case 1 in the list above,
+but I have had requests to support 2 and 3.]
+
+When performing type expansion for an argument, the argument that
+is targeted for expansion should be evaluated without the use of
+any context. All arguments that are not yet expanded should
+continue to be evaluated with the benefit of context supplied by parameter
+types within each overload signature.
+
+Example::
+
+  class MyDict[T](TypedDict):
+      x: T
+
+  @overload
+  def func[T](a: int, b: MyDict[T]) -> T: ...
+
+  @overload
+  def func(a: str, b: dict[str, int]) -> str: ...
+
+
+  def test(val: int | str):
+      result = func(val, {'x': 1})
+      reveal_type(result)  # Should reveal "int | str"
+
+In this case, type expansion is performed on the first argument,
+which expands its type from ``int | str`` to ``int`` and ``str``.
+The expression for the second argument is evaluated in the context
+of both overloads. For the first overload, the second argument evaluates
+to ``MyDict[int]``, and for the second overload it evaluates to
+``dict[str, int]``. Both overloads are used to evaluate this call,
+and the final type of ``result`` is ``int | str``.
+
+[Eric's note: mypy apparently doesn't do this currently. It evaluates all
+arguments without the benefit of context, which produces less-than-ideal
+results in some cases.]
+
+
+[Eric's note for reviewers: We may want to provide for argument type expansion
+for regular (non-overloaded) calls as well. This came up recently in
+[this thread](https://discuss.python.org/t/proposal-relax-un-correlated-constrained-typevars/59658).
+I'm a bit hesitant to add this to the spec because it adds significant
+complexity to call evaluations and would likely result in a measurable slowdown
+in type evaluation, but it's worth considering. We could perhaps mitigate the
+slowdown by applying this behavior only when a constrained type variable is
+used in the call's signature.]
```

Comment:
I think that we should support this for constrained TypeVars. As I noted above, the spec already says unequivocally that a constrained TypeVar is equivalent to an overload, and failing to expand unions for constrained TypeVars violates that equivalence.

I'm not sure I see where this expansion could ever be meaningful outside of overloads or constrained TypeVars? With a bound TypeVar, it would require a case where the union type of the argument is not a subtype of the bound, but the individual elements of the union are all subtypes of the bound? That seems impossible, as it would violate the definition of union subtyping.

Reply:
I wish I could also go back in time and prevent the introduction of constrained TypeVars. There's a good reason they don't appear anywhere in other programming languages. 

Doing argument expansion roughly triples the cost of evaluating a call expression. If there are nested call expressions (e.g. a call expression used within an argument expression for another call), these costs multiply. It's not uncommon in scientific code to have six or more nested layers. That could mean a 700x or more performance penalty just to support an edge case that almost never comes up in practice. I'm sensitive to this because I sometimes see this with nested calls that invoke overloads, and the performance impact can be huge. Pyright has a bunch of smarts to try to mitigate this overhead, but there's only so much that can be done. There are cases where evaluating one expression can take many seconds, which is really bad if the analyzer is powering a language server. (And no, implementing the type checker in Rust will not solve this problem. ;) Mypy also chokes on deeply-nested calls. I want to be careful about changing the spec to mandate support for some esoteric edge case if it's going to mean that millions of developers will be negatively impacted.

In any case, I don't think we need to decide this now for regular (non-overloaded) functions since this PR is focused on clarifications for the overload chapter. We can consider such a change in a future update.

Reply:
> implementing the type checker in Rust will not solve this problem. ;)

Little-known fact: if you use Rust, all algorithms become O(1). True story ;)

The concern here makes sense. I'm ok deferring this to later consideration of constrained TypeVars. My argument would be that we already have this potential performance problem, with overloads, and I'm not sure that drawing what seems like an arbitrary line between overloads and constrained TypeVars (when, as the spec already clearly says, use of a constrained TypeVar should be equivalent to an overload) is the best way to address the performance issue. I would rather have the spec be internally consistent and correctly reflect the type-level equivalence, and then also consider holistically how the spec can allow type checkers to soften the potential performance impact of deeply nested union expansions, whether they occur with overloads or with TypeVars.

Put another way: it's also not exactly nice if changing a constrained TypeVar to the equivalent overload means that now your type checker hangs where it didn't before.